e2e/capabilities/workflow_test.go

```diff
@@ -0,0 +1,521 @@
+		bc, err := blockchain.NewBlockchainNetwork(in.BlockchainA)
+		require.NoError(t, err)
+
+		// connect clients
+		sc, err := seth.NewClientBuilder().
+			WithRpcUrl(bc.Nodes[0].HostWSUrl).
+			WithPrivateKeys([]string{pkey}).
+			Build()
+		require.NoError(t, err)
+
+		capabilitiesRegistryAddress, tx, _, err := capabilities_registry.DeployCapabilitiesRegistry(
+			sc.NewTXOpts(),
+			sc.Client,
+		)
+		require.NoError(t, err)
+		_, err = bind.WaitMined(context.Background(), sc.Client, tx)
+		require.NoError(t, err)
+		fmt.Println("Deployed capabilities_registry contract at", capabilitiesRegistryAddress)
+
+		forwarderAddress, tx, _, err := forwarder.DeployKeystoneForwarder(
+			sc.NewTXOpts(),
+			sc.Client,
+		)
+		require.NoError(t, err)
+		_, err = bind.WaitMined(context.Background(), sc.Client, tx)
+		require.NoError(t, err)
+		fmt.Println("Deployed forwarder contract at", forwarderAddress)
+
+		for _, n := range in.NodeSet.NodeSpecs {
+			n.Node.UserConfigOverrides = fmt.Sprintf(`
+			[Feature]
+			LogPoller = true
+
+			[OCR2]
+			Enabled = true
+			DatabaseTimeout = '1s'
+
+			[P2P.V2]
+			Enabled = true
+			ListenAddresses = ['0.0.0.0:6690']
+
+			# This is needed for the target capability to be initialized
+			[[EVM]]
+			ChainID = '%s'
+
+			[[EVM.Nodes]]
+			Name = 'anvil'
+			WSURL = '%s'
+			HTTPURL = '%s'
+			`,
+				bc.ChainID,
+				bc.Nodes[0].HostWSUrl,
+				bc.Nodes[0].HostHTTPUrl,
+			)
+		}
+
+		// TODO: When the capabilities registry address is provided:
+		// - NOPs and nodes are added to the registry.
+		// - Nodes are configured to listen to the registry for updates.
+		nodeset, err := ns.NewSharedDBNodeSet(in.NodeSet, bc, "https://example.com") // TODO: Should not be a thing
+		require.NoError(t, err)
+
+		for i, n := range nodeset.CLNodes {
+			fmt.Printf("Node %d --> %s\n", i, n.Node.HostURL)
+			fmt.Printf("Node P2P %d --> %s\n", i, n.Node.HostP2PURL)
+		}
+
+		nodeClients, err := clclient.NewCLDefaultClients(nodeset.CLNodes, framework.L)
+		require.NoError(t, err)
+		nodesInfo := getNodesInfo(t, nodeClients)
+
+		for i, node := range nodeClients {
+			fmt.Println("Node i ", i)
+			fmt.Println("Node ", node)
+			// First node is a bootstrap node, so we skip it
+			if i == 0 {
+				continue
+			}
+
+			in.NodeSet.NodeSpecs[i].Node.UserConfigOverrides = fmt.Sprintf(`
+				[Feature]
+				LogPoller = true
+
+				[OCR2]
+				Enabled = true
+				DatabaseTimeout = '1s'
+
+				[P2P.V2]
+				Enabled = true
+				ListenAddresses = ['0.0.0.0:6690']
+
+				# This is needed for the target capability to be initialized
+				[[EVM]]
+				ChainID = '%s'
+
+				[[EVM.Nodes]]
+				Name = 'anvil'
+				WSURL = '%s'
+				HTTPURL = '%s'
+
+				[EVM.Workflow]
+				FromAddress = '%s'
+				ForwarderAddress = '%s'
+				GasLimitDefault = 400_000
+
+				# This is needed for external registry
+				[Capabilities.ExternalRegistry]
+				Address = '%s'
+				NetworkID = 'evm'
+				ChainID = '%s'
+			`,
+				bc.ChainID,
+				bc.Nodes[0].HostWSUrl,
```

Comment:
should be using Docker URL instead since communication is performed inside Docker network